File: typescript/packages/mechanisms/evm/src/erc4337/bundler.ts
/**
 * ERC-4337 Bundler Client
 *
 * Client for interacting with ERC-4337 bundlers via JSON-RPC.
 * Handles UserOperation submission, gas estimation, and receipt polling.
 */

import type { Address, Hex } from "viem";
import type {
  UserOperation,
  PackedUserOperation,
  GasEstimate,
  UserOperationReceipt,
  UserOperationResult,
  BundlerConfig,
} from "./types.js";
import {
  ENTRYPOINT_V07_ADDRESS,
  BUNDLER_METHODS,
  packAccountGasLimits,
  packGasFees,
} from "./constants.js";

/**
 * JSON-RPC request structure
 */
interface JsonRpcRequest {
  jsonrpc: "2.0";
  id: number;
  method: string;
  params: unknown[];
}

/**
 * JSON-RPC response structure
 */
interface JsonRpcResponse<T = unknown> {
  jsonrpc: "2.0";
  id: number;
  result?: T;
  error?: {
    code: number;
    message: string;
    data?: unknown;
  };
}

/**
 * Bundler error class
 */
export class BundlerError extends Error {
  constructor(
    message: string,
    public code?: number,
    public data?: unknown,
  ) {
    super(message);
    this.name = "BundlerError";
  }
}

/**
 * Bundler client for submitting UserOperations
 */
export class BundlerClient {
  private readonly bundlerUrl: string;
  private readonly entryPoint: Address;
  private readonly chainId: number;
  private requestId: number = 0;

  constructor(config: BundlerConfig) {
    this.bundlerUrl = config.bundlerUrl;
    this.entryPoint = config.entryPoint ?? ENTRYPOINT_V07_ADDRESS;
    this.chainId = config.chainId;
  }

  /**
   * Send a UserOperation to the bundler
   */
  async sendUserOperation(userOp: UserOperation): Promise<UserOperationResult> {
    const packed = this.packForRpc(userOp);

    const userOpHash = await this.rpcCall<Hex>(
      BUNDLER_METHODS.sendUserOperation,
      [packed, this.entryPoint],
    );

    return {
      userOpHash,
      wait: () => this.waitForReceipt(userOpHash),
    };
  }

  /**
   * Estimate gas for a UserOperation
   */
  async estimateUserOperationGas(
    userOp: Partial<UserOperation> & {
      sender: Address;
      callData: Hex;
    },
  ): Promise<GasEstimate> {
    // Fill in defaults for estimation
    const estimationOp = {
      sender: userOp.sender,
      nonce: userOp.nonce ?? 0n,
      initCode: userOp.initCode ?? "0x",
      callData: userOp.callData,
      verificationGasLimit: userOp.verificationGasLimit ?? 1000000n,
      callGasLimit: userOp.callGasLimit ?? 1000000n,
      preVerificationGas: userOp.preVerificationGas ?? 100000n,
      maxPriorityFeePerGas: userOp.maxPriorityFeePerGas ?? 1000000000n,
      maxFeePerGas: userOp.maxFeePerGas ?? 10000000000n,
      paymasterAndData: userOp.paymasterAndData ?? "0x",
      signature:
        userOp.signature ??
        // Dummy signature for estimation
        "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c",
    };

    const packed = this.packForRpc(estimationOp as UserOperation);

    const result = await this.rpcCall<{
      verificationGasLimit: Hex;
      callGasLimit: Hex;
      preVerificationGas: Hex;
      paymasterVerificationGasLimit?: Hex;
      paymasterPostOpGasLimit?: Hex;
    }>(BUNDLER_METHODS.estimateUserOperationGas, [packed, this.entryPoint]);

    return {
      verificationGasLimit: BigInt(result.verificationGasLimit),
      callGasLimit: BigInt(result.callGasLimit),
      preVerificationGas: BigInt(result.preVerificationGas),
      paymasterVerificationGasLimit: result.paymasterVerificationGasLimit
        ? BigInt(result.paymasterVerificationGasLimit)
        : undefined,
      paymasterPostOpGasLimit: result.paymasterPostOpGasLimit
        ? BigInt(result.paymasterPostOpGasLimit)
        : undefined,
    };
  }

  /**
   * Get UserOperation by hash
   */
  async getUserOperationByHash(
    userOpHash: Hex,
  ): Promise<{ userOperation: PackedUserOperation; entryPoint: Address } | null> {
    const result = await this.rpcCall<{
      userOperation: PackedUserOperation;
      entryPoint: Address;
    } | null>(BUNDLER_METHODS.getUserOperationByHash, [userOpHash]);

    return result;
  }

  /**
   * Get UserOperation receipt
   */
  async getUserOperationReceipt(
    userOpHash: Hex,
  ): Promise<UserOperationReceipt | null> {
    const result = await this.rpcCall<{
      userOpHash: Hex;
      sender: Address;
      nonce: Hex;
      paymaster?: Address;
      actualGasCost: Hex;
      actualGasUsed: Hex;
      success: boolean;
      reason?: string;
      receipt: {
        transactionHash: Hex;
        blockNumber: Hex;
        blockHash: Hex;
      };
    } | null>(BUNDLER_METHODS.getUserOperationReceipt, [userOpHash]);

    if (!result) return null;

    return {
      userOpHash: result.userOpHash,
      sender: result.sender,
      nonce: BigInt(result.nonce),
      paymaster: result.paymaster,
      actualGasCost: BigInt(result.actualGasCost),
      actualGasUsed: BigInt(result.actualGasUsed),
      success: result.success,
      reason: result.reason,
      receipt: {
        transactionHash: result.receipt.transactionHash,
        blockNumber: BigInt(result.receipt.blockNumber),
        blockHash: result.receipt.blockHash,
      },
    };
  }

  /**
   * Get supported EntryPoints
   */
  async getSupportedEntryPoints(): Promise<Address[]> {
    return this.rpcCall<Address[]>(BUNDLER_METHODS.supportedEntryPoints, []);
  }

  /**
   * Get chain ID from bundler
   */
  async getChainId(): Promise<number> {
    const result = await this.rpcCall<Hex>(BUNDLER_METHODS.chainId, []);
    return Number(result);
  }

  /**
   * Wait for UserOperation receipt with polling
   */
  async waitForReceipt(
    userOpHash: Hex,
    options: { timeout?: number; pollingInterval?: number } = {},
  ): Promise<UserOperationReceipt> {
    const timeout = options.timeout ?? 60000; // 60 seconds default
    const pollingInterval = options.pollingInterval ?? 2000; // 2 seconds default

    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const receipt = await this.getUserOperationReceipt(userOpHash);

      if (receipt) {
        return receipt;
      }

      await new Promise((resolve) => setTimeout(resolve, pollingInterval));
    }

    throw new BundlerError(
      `Timeout waiting for UserOperation receipt: ${userOpHash}`,
    );
  }

  /**
   * Pack UserOperation for RPC (convert bigints to hex strings)
   */
  private packForRpc(userOp: UserOperation): Record<string, unknown> {
    return {
      sender: userOp.sender,
      nonce: this.toHex(userOp.nonce),
      initCode: userOp.initCode,
      callData: userOp.callData,
      accountGasLimits: packAccountGasLimits(
        userOp.verificationGasLimit,
        userOp.callGasLimit,
      ),
      preVerificationGas: this.toHex(userOp.preVerificationGas),
      gasFees: packGasFees(userOp.maxPriorityFeePerGas, userOp.maxFeePerGas),
      paymasterAndData: userOp.paymasterAndData,
      signature: userOp.signature,
    };
  }

  /**
   * Convert bigint to hex string
   */
  private toHex(value: bigint): Hex {
    return `0x${value.toString(16)}` as Hex;
  }

  /**
   * Make a JSON-RPC call to the bundler
   */
  private async rpcCall<T>(method: string, params: unknown[]): Promise<T> {
    const request: JsonRpcRequest = {
      jsonrpc: "2.0",
      id: ++this.requestId,
      method,
      params,
    };

    const response = await fetch(this.bundlerUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      throw new BundlerError(
        `HTTP error: ${response.status} ${response.statusText}`,
      );
    }

    const json = (await response.json()) as JsonRpcResponse<T>;

    if (json.error) {
      throw new BundlerError(json.error.message, json.error.code, json.error.data);
    }

    return json.result as T;
  }
}

/**
 * Create a BundlerClient instance
 */
export function createBundlerClient(config: BundlerConfig): BundlerClient {
  return new BundlerClient(config);
}

File: go/mechanisms/evm/erc4337/bundler.go
package erc4337

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"strings"
	"time"

	"github.com/ethereum/go-ethereum/common"
)

// GenericBundlerClient is a generic bundler client implementation.
type GenericBundlerClient struct {
	bundlerURL string
	entryPoint common.Address
	chainID    int64
	requestID  int
	httpClient *http.Client
}

// NewBundlerClient creates a new generic bundler client.
func NewBundlerClient(config BundlerConfig) *GenericBundlerClient {
	entryPoint := config.EntryPoint
	if entryPoint == (common.Address{}) {
		entryPoint = common.HexToAddress(EntryPointV07Address)
	}

	return &GenericBundlerClient{
		bundlerURL: config.BundlerURL,
		entryPoint: entryPoint,
		chainID:    config.ChainID,
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}
}

// SendUserOperation submits a UserOperation to the bundler.
func (c *GenericBundlerClient) SendUserOperation(userOp *UserOperation) (common.Hash, error) {
	packed := c.packUserOp(userOp)

	var result string
	err := c.rpcCall(BundlerMethods.SendUserOperation, []interface{}{packed, c.entryPoint.Hex()}, &result)
	if err != nil {
		return common.Hash{}, err
	}

	return common.HexToHash(result), nil
}

// EstimateUserOperationGas estimates gas for a UserOperation.
func (c *GenericBundlerClient) EstimateUserOperationGas(userOp *UserOperation) (*GasEstimate, error) {
	packed := c.packUserOp(userOp)

	var result struct {
		VerificationGasLimit          string `json:"verificationGasLimit"`
		CallGasLimit                  string `json:"callGasLimit"`
		PreVerificationGas            string `json:"preVerificationGas"`
		PaymasterVerificationGasLimit string `json:"paymasterVerificationGasLimit,omitempty"`
		PaymasterPostOpGasLimit       string `json:"paymasterPostOpGasLimit,omitempty"`
	}

	err := c.rpcCall(BundlerMethods.EstimateUserOperationGas, []interface{}{packed, c.entryPoint.Hex()}, &result)
	if err != nil {
		return nil, err
	}

	estimate := &GasEstimate{
		VerificationGasLimit: hexToBigInt(result.VerificationGasLimit),
		CallGasLimit:         hexToBigInt(result.CallGasLimit),
		PreVerificationGas:   hexToBigInt(result.PreVerificationGas),
	}

	if result.PaymasterVerificationGasLimit != "" {
		estimate.PaymasterVerificationGasLimit = hexToBigInt(result.PaymasterVerificationGasLimit)
	}
	if result.PaymasterPostOpGasLimit != "" {
		estimate.PaymasterPostOpGasLimit = hexToBigInt(result.PaymasterPostOpGasLimit)
	}

	return estimate, nil
}

// GetUserOperationByHash retrieves a UserOperation by hash.
func (c *GenericBundlerClient) GetUserOperationByHash(hash common.Hash) (*UserOperation, error) {
	var result struct {
		UserOperation map[string]interface{} `json:"userOperation"`
		EntryPoint    string                 `json:"entryPoint"`
	}

	err := c.rpcCall(BundlerMethods.GetUserOperationByHash, []interface{}{hash.Hex()}, &result)
	if err != nil {
		return nil, err
	}

	if result.UserOperation == nil {
		return nil, nil
	}

	// Parse the UserOperation from the result
	return parseUserOp(result.UserOperation), nil
}

// GetUserOperationReceipt retrieves the receipt for a UserOperation.
func (c *GenericBundlerClient) GetUserOperationReceipt(hash common.Hash) (*UserOperationReceipt, error) {
	var result struct {
		UserOpHash    string `json:"userOpHash"`
		Sender        string `json:"sender"`
		Nonce         string `json:"nonce"`
		Paymaster     string `json:"paymaster,omitempty"`
		ActualGasCost string `json:"actualGasCost"`
		ActualGasUsed string `json:"actualGasUsed"`
		Success       bool   `json:"success"`
		Reason        string `json:"reason,omitempty"`
		Receipt       struct {
			TransactionHash string `json:"transactionHash"`
			BlockNumber     string `json:"blockNumber"`
			BlockHash       string `json:"blockHash"`
		} `json:"receipt"`
	}

	err := c.rpcCall(BundlerMethods.GetUserOperationReceipt, []interface{}{hash.Hex()}, &result)
	if err != nil {
		return nil, err
	}

	if result.UserOpHash == "" {
		return nil, nil
	}

	receipt := &UserOperationReceipt{
		UserOpHash:    common.HexToHash(result.UserOpHash),
		Sender:        common.HexToAddress(result.Sender),
		Nonce:         hexToBigInt(result.Nonce),
		ActualGasCost: hexToBigInt(result.ActualGasCost),
		ActualGasUsed: hexToBigInt(result.ActualGasUsed),
		Success:       result.Success,
		Reason:        result.Reason,
		Receipt: TransactionReceipt{
			TransactionHash: common.HexToHash(result.Receipt.TransactionHash),
			BlockNumber:     hexToBigInt(result.Receipt.BlockNumber),
			BlockHash:       common.HexToHash(result.Receipt.BlockHash),
		},
	}

	if result.Paymaster != "" && result.Paymaster != "0x" {
		paymaster := common.HexToAddress(result.Paymaster)
		receipt.Paymaster = &paymaster
	}

	return receipt, nil
}

// GetSupportedEntryPoints returns supported EntryPoint addresses.
func (c *GenericBundlerClient) GetSupportedEntryPoints() ([]common.Address, error) {
	var result []string
	err := c.rpcCall(BundlerMethods.SupportedEntryPoints, []interface{}{}, &result)
	if err != nil {
		return nil, err
	}

	addresses := make([]common.Address, len(result))
	for i, addr := range result {
		addresses[i] = common.HexToAddress(addr)
	}

	return addresses, nil
}

// WaitForReceipt waits for a UserOperation receipt with polling.
func (c *GenericBundlerClient) WaitForReceipt(hash common.Hash, timeout, pollingInterval time.Duration) (*UserOperationReceipt, error) {
	if timeout == 0 {
		timeout = 60 * time.Second
	}
	if pollingInterval == 0 {
		pollingInterval = 2 * time.Second
	}

	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		receipt, err := c.GetUserOperationReceipt(hash)
		if err != nil {
			return nil, err
		}
		if receipt != nil {
			return receipt, nil
		}
		time.Sleep(pollingInterval)
	}

	return nil, fmt.Errorf("timeout waiting for UserOperation receipt: %s", hash.Hex())
}

// packUserOp packs a UserOperation for RPC transmission.
func (c *GenericBundlerClient) packUserOp(userOp *UserOperation) map[string]interface{} {
	accountGasLimits := PackAccountGasLimits(userOp.VerificationGasLimit, userOp.CallGasLimit)
	gasFees := PackGasFees(userOp.MaxPriorityFeePerGas, userOp.MaxFeePerGas)

	return map[string]interface{}{
		"sender":             userOp.Sender.Hex(),
		"nonce":              bigIntToHex(userOp.Nonce),
		"initCode":           bytesToHex(userOp.InitCode),
		"callData":           bytesToHex(userOp.CallData),
		"accountGasLimits":   bytesToHex(accountGasLimits[:]),
		"preVerificationGas": bigIntToHex(userOp.PreVerificationGas),
		"gasFees":            bytesToHex(gasFees[:]),
		"paymasterAndData":   bytesToHex(userOp.PaymasterAndData),
		"signature":          bytesToHex(userOp.Signature),
	}
}

// rpcCall makes a JSON-RPC call to the bundler.
func (c *GenericBundlerClient) rpcCall(method string, params []interface{}, result interface{}) error {
	c.requestID++

	request := map[string]interface{}{
		"jsonrpc": "2.0",
		"id":      c.requestID,
		"method":  method,
		"params":  params,
	}

	body, err := json.Marshal(request)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	resp, err := c.httpClient.Post(c.bundlerURL, "application/json", bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("HTTP request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("HTTP error %d: %s", resp.StatusCode, string(body))
	}

	var response struct {
		Result json.RawMessage `json:"result"`
		Error  *struct {
			Code    int             `json:"code"`
			Message string          `json:"message"`
			Data    json.RawMessage `json:"data,omitempty"`
		} `json:"error"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return fmt.Errorf("failed to decode response: %w", err)
	}

	if response.Error != nil {
		return fmt.Errorf("RPC error %d: %s", response.Error.Code, response.Error.Message)
	}

	if result != nil && len(response.Result) > 0 {
		if err := json.Unmarshal(response.Result, result); err != nil {
			return fmt.Errorf("failed to unmarshal result: %w", err)
		}
	}

	return nil
}

// Helper functions

func bigIntToHex(n *big.Int) string {
	if n == nil {
		return "0x0"
	}
	return "0x" + n.Text(16)
}

func bytesToHex(b []byte) string {
	if len(b) == 0 {
		return "0x"
	}
	return "0x" + hex.EncodeToString(b)
}

func hexToBigInt(s string) *big.Int {
	s = strings.TrimPrefix(s, "0x")
	if s == "" {
		return big.NewInt(0)
	}
	n := new(big.Int)
	n.SetString(s, 16)
	return n
}

func hexToBytes(s string) []byte {
	s = strings.TrimPrefix(s, "0x")
	if s == "" {
		return nil
	}
	b, _ := hex.DecodeString(s)
	return b
}

func parseUserOp(data map[string]interface{}) *UserOperation {
	userOp := &UserOperation{}

	if sender, ok := data["sender"].(string); ok {
		userOp.Sender = common.HexToAddress(sender)
	}
	if nonce, ok := data["nonce"].(string); ok {
		userOp.Nonce = hexToBigInt(nonce)
	}
	if initCode, ok := data["initCode"].(string); ok {
		userOp.InitCode = hexToBytes(initCode)
	}
	if callData, ok := data["callData"].(string); ok {
		userOp.CallData = hexToBytes(callData)
	}
	if verificationGasLimit, ok := data["verificationGasLimit"].(string); ok {
		userOp.VerificationGasLimit = hexToBigInt(verificationGasLimit)
	}
	if callGasLimit, ok := data["callGasLimit"].(string); ok {
		userOp.CallGasLimit = hexToBigInt(callGasLimit)
	}
	if preVerificationGas, ok := data["preVerificationGas"].(string); ok {
		userOp.PreVerificationGas = hexToBigInt(preVerificationGas)
	}
	if maxPriorityFeePerGas, ok := data["maxPriorityFeePerGas"].(string); ok {
		userOp.MaxPriorityFeePerGas = hexToBigInt(maxPriorityFeePerGas)
	}
	if maxFeePerGas, ok := data["maxFeePerGas"].(string); ok {
		userOp.MaxFeePerGas = hexToBigInt(maxFeePerGas)
	}
	if paymasterAndData, ok := data["paymasterAndData"].(string); ok {
		userOp.PaymasterAndData = hexToBytes(paymasterAndData)
	}
	if signature, ok := data["signature"].(string); ok {
		userOp.Signature = hexToBytes(signature)
	}

	return userOp
}
